pkg/nsx/services/common/policy_tree.go

```diff
@@ -0,0 +1,535 @@
+			continue
+		}
+		rootNode.mergeChildNode(orgNode, b.leafType)
+	}
+	return rootNode
+}
+
+func (b *PolicyTreeBuilder[T]) buildTree(resources []T, parentPath string) ([]*data.StructValue, error) {
+	rootNode := b.BuildRootNode(resources, parentPath)
+	children, err := rootNode.buildTree(b.rootType, b.leafType)
+	if err != nil {
+		log.Error(err, "Failed to build data values for multiple resources")
+		return nil, err
+	}
+	return children, nil
+}
+
+func (b *PolicyTreeBuilder[T]) BuildOrgRoot(resources []T, parentPath string) (*model.OrgRoot, error) {
+	children, err := b.buildTree(resources, parentPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return &model.OrgRoot{
+		Children:     children,
+		ResourceType: String(ResourceTypeOrgRoot),
+	}, nil
+}
+
+func (b *PolicyTreeBuilder[T]) BuildInfra(resources []T, parentPath string) (*model.Infra, error) {
+	children, err := b.buildTree(resources, parentPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return wrapInfra(children), nil
+}
+
+func (b *PolicyTreeBuilder[T]) buildHNodeFromResource(path string, res T) (*hNode[T], error) {
+	pathSegments, err := b.parsePathSegments(path)
+	if err != nil {
+		return nil, err
+	}
+
+	dataValue, err := b.leafWrapper(res)
+	if err != nil {
+		return nil, err
+	}
+
+	idx := len(pathSegments) - 1
+	nodeCount := len(b.pathFormat)
+	nodes := make([]*hNode[T], nodeCount)
+	leafIdx := nodeCount - 1
+	nodes[leafIdx] = &hNode[T]{
+		key: &hNodeKey{
+			resID:   pathSegments[idx],
+			resType: b.modelFormat[leafIdx],
+		},
+		leafData: dataValue,
+	}
+	idx -= 2
+
+	for i := leafIdx - 1; i >= 0; i-- {
+		child := nodes[i+1]
+		resType := b.modelFormat[i]
+		var resID string
+		if resType != ResourceTypeInfra {
+			resID = pathSegments[idx]
+			idx -= 2
+		} else {
+			resID = ""
+			idx -= 1
+		}
+		node := &hNode[T]{
+			key: &hNodeKey{
+				resID:   resID,
+				resType: resType,
+			},
+			childNodes: map[hNodeKey]*hNode[T]{
+				*child.key: child,
+			},
+		}
+		nodes[i] = node
+	}
+	n := nodes[0]
+	return n, nil
+}
+
+func (b *PolicyTreeBuilder[T]) parsePathSegments(inputPath string) ([]string, error) {
+	pathSegments := strings.Split(strings.Trim(inputPath, "/"), "/")
+	// Remove "infra" in the path since the infra resource's path does not follow the format "/key/value", so we remove
+	// the first "infra" in the pathSegments.
+	if b.rootType == ResourceTypeInfra {
+		pathSegments = pathSegments[1:]
+	}
+
+	segmentsCount := len(pathSegments)
+	leafPathKey := b.pathFormat[len(b.pathFormat)-1]
+	if segmentsCount < 2 || pathSegments[segmentsCount-2] != leafPathKey {
+		return nil, fmt.Errorf("invalid input path %s for resource %s", inputPath, b.leafType)
+	}
+	if b.hasInnerInfra {
+		segmentsCount += 1
+	}
+	if segmentsCount != len(b.pathFormat)*2 {
+		return nil, fmt.Errorf("invalid input path: %s", inputPath)
+	}
+
+	return pathSegments, nil
+}
+
+func (b *PolicyTreeBuilder[T]) DeleteMultipleResourcesOnNSX(objects []T, nsxClient *nsx.Client) error {
```

Comment:
The function name is a little bit confusing as it looks like the function is for update instead of deletion. It deletes the objects only if the caller sets markForDelete to true. Would you think it's better to rename to UpdateMultipleResourcesOnNSX or add some comments to clarify this?
If you choose to rename the function, maybe also update PagingDeleteResources?